### PR TITLE
Fix/merge cubes

### PR DESCRIPTION
**Pull request by Ardweaden**
Closes https://git.sinergise.com/team-6/openeo-platform/-/issues/105

### CODE REVIEW

File: src/pg_to_evalscript/javascript_datacube/DataCube.js [hunk 1]
```diff
@@ -443,4 +495,172 @@ class DataCube {
...
+                name: "OverlapResolverMissing",
+                message: "Overlapping data cubes, but no overlap resolver has been specified."
+            });
+        }
+
+        if (allDimensionsEqual && !dimensionWithDifferentLabels) {
+            return this._merge_matching_cube(cube2, overlap_resolver)
+        }
+
+        const isCube2Subcube = !dimensionWithDifferentLabels && cube2SpecificDimensions.length === 0 && cube1SpecificDimensions.length > 0;
```
Comment (dgostencnik) [inline src/pg_to_evalscript/javascript_datacube/DataCube.js:647]: Can it also be the other way around and cube1 is subcube of cube2 and low dimension cube (cube1) is being merged with higher dimension cube(cube2)? Cube1 has spatial dimension with x label, cube2 has spatial dimension with x and y.

Most likely this is (also) handled by code below.

File: src/pg_to_evalscript/javascript_datacube/DataCube.js [hunk 2]
```diff
@@ -270,24 +270,73 @@ class DataCube {
...
             );
         }
 
         return dataArr;
     }
 
+    // axis: integer, dimension axis
+    // dataToAdd: array, values to insert in dimension.
+    // locationInDimension: integer, offset from start of dimension, max length of dimension. Default 0.
+    insertIntoDimension(axis, dataToAdd, locationInDimension = 0) {
```
Comment (dgostencnik) [inline src/pg_to_evalscript/javascript_datacube/DataCube.js:294]: it would probably make sense to add tests for newly added functions.fftr

Comment (Ardweaden) [reply src/pg_to_evalscript/javascript_datacube/DataCube.js:394]: Added tests for `insertIntoDimension` and `setInDimension`.

File: src/pg_to_evalscript/javascript_datacube/DataCube.js (hunk 1 above)
Comment (Ardweaden) [reply src/pg_to_evalscript/javascript_datacube/DataCube.js:781]: Nice spot, it doesn't work, added a test and changes to handle the opposite order.

File: src/pg_to_evalscript/javascript_datacube/DataCube.js [hunk 3]
```diff
@@ -443,4 +495,172 @@ class DataCube {
...
+        let dimensionWithDifferentLabelsOverlaps = false;
+
+        for (let dimension of this.dimensions) {
+            const dimension2 = cube2.getDimensionByName(dimension.name);
+
+            if (!dimension2) {
+                cube1SpecificDimensions.push(dimension)
+                continue
+            } else {
+                sharedDimensions.push(dimension)
```
[the remark this answers is not in the document]
Comment (Ardweaden) [reply src/pg_to_evalscript/javascript_datacube/DataCube.js]: Removed.

File: src/pg_to_evalscript/javascript_datacube/DataCube.js [hunk 4]
```diff
@@ -443,4 +495,172 @@ class DataCube {
...
+        for (let dimension2 of cube2.dimensions) {
+            const dimension = this.getDimensionByName(dimension2.name);
+            if (!dimension) {
+                cube2SpecificDimensions.push(dimension2)
+            }
+        }
+
+        const allDimensionsEqual = cube1SpecificDimensions.length === 0 && cube2SpecificDimensions.length === 0;
+
+        if (!overlap_resolver && ((dimensionWithDifferentLabels && dimensionWithDifferentLabelsOverlaps) || (!dimensionWithDifferentLabels && allDimensionsEqual))) {
```
[the remark this answers is not in the document]
Comment (Ardweaden) [reply src/pg_to_evalscript/javascript_datacube/DataCube.js]: Thanks. I misunderstood it as if the the backend can decide which behaviour to implement, but it indeed makes more sense to just handle both cases. Should work now.

File: src/pg_to_evalscript/javascript_datacube/DataCube.js [hunk 5]
```diff
@@ -443,4 +495,172 @@ class DataCube {
...
+            const dimension2 = cube2.getDimensionByName(dimension.name);
+
+            if (!dimension2) {
+                cube1SpecificDimensions.push(dimension)
+                continue
+            } else {
+                sharedDimensions.push(dimension)
+            }
+
+            const labelsEqual = dimension.labels.length === dimension2.labels.length && dimension.labels.every((l) => dimension2.labels.includes(l))
```
[the remark this answers is not in the document]
Comment (Ardweaden) [reply src/pg_to_evalscript/javascript_datacube/DataCube.js]: I don't think so, at least it doesn't make much sense. Something must've gone wrong if there are duplicated layers. Of course in the example above the first datacube should have data `[1, 1, 3, 3, 5, 5]` or something, as have different values for the same band at the same time surely violates some principles....

Anyway, I decided to just raise and error if the labels are duplicated.

**Review by dgostencnik — APPROVED** [review]

File: src/pg_to_evalscript/javascript_datacube/DataCube.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ardweaden) [reply src/pg_to_evalscript/javascript_datacube/DataCube.js]: This test looks identical to the test that was already added when fixing this: https://github.com/openEOPlatform/openeo-pg-evalscript-converter/pull/71/commits/3d617d42b70d47695ea818b0e2c88ef67164b275

[the remark this answers is not in the document]
Comment (dgostencnik) [reply src/pg_to_evalscript/javascript_datacube/DataCube.js]: lol, I'm blind

**Pull request merged**